File: apps/macos/Kelpie/Handlers/TapCalibrationHandler.swift
import Foundation

struct TapCalibrationHandler {
    func register(on router: Router) {
        router.register("get-tap-calibration") { _ in getCalibration() }
        router.register("set-tap-calibration") { body in setCalibration(body) }
    }

    private func getCalibration() -> [String: Any] {
        successResponse(TapCalibrationStore.current().responsePayload)
    }

    private func setCalibration(_ body: [String: Any]) -> [String: Any] {
        guard let offsetX = double(body["offsetX"]),
              let offsetY = double(body["offsetY"]),
              offsetX.isFinite,
              offsetY.isFinite else {
            return errorResponse(
                code: "MISSING_PARAM",
                message: "offsetX and offsetY are required numbers"
            )
        }
        return successResponse(TapCalibrationStore.save(offsetX: offsetX, offsetY: offsetY).responsePayload)
    }

    private func double(_ value: Any?) -> Double? {
        (value as? NSNumber)?.doubleValue
    }
}

File: apps/android/app/src/main/java/com/kelpie/browser/handlers/TapCalibrationHandler.kt
package com.kelpie.browser.handlers

import com.kelpie.browser.network.Router
import com.kelpie.browser.network.errorResponse
import com.kelpie.browser.network.successResponse

class TapCalibrationHandler {
    fun register(router: Router) {
        router.register("get-tap-calibration") { getCalibration() }
        router.register("set-tap-calibration") { body -> setCalibration(body) }
    }

    private fun getCalibration() =
        successResponse(
            mapOf(
                "offsetX" to TapCalibrationStore.current().offsetX,
                "offsetY" to TapCalibrationStore.current().offsetY,
            ),
        )

    private fun setCalibration(body: Map<String, Any?>): Map<String, Any?> {
        val offsetX = (body["offsetX"] as? Number)?.toDouble()
        val offsetY = (body["offsetY"] as? Number)?.toDouble()
        if (offsetX == null || offsetY == null || !offsetX.isFinite() || !offsetY.isFinite()) {
            return errorResponse("MISSING_PARAM", "offsetX and offsetY are required numbers")
        }
        val saved = TapCalibrationStore.save(offsetX, offsetY)
        return successResponse(
            mapOf(
                "offsetX" to saved.offsetX,
                "offsetY" to saved.offsetY,
            ),
        )
    }
}
